### PR TITLE
Partial safety comments

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on https://github.com/hsivonen/encoding_rs/issues/98

Figured I'd land what I have so far. 

I still need to do:

 - mem.rs (partial)
 - lib.rs
 - utf_8.rs

Furthermore, I've not done too much commenting in handles.rs: I think there are some refactors to that code that may reduce the number of tricky invariants.

There's also a safety todo around endianness in ascii.rs.

The general methodology I'm using is:

 - document every unsafe block
 - if the unsafe block relies on invariants that are not immediately checked above it, document that invariant on the relevant variable (and if it is mutated, on every mutation point)
 - make sure all functions have safety invariants documented

### CODE REVIEW

File: src/single_byte.rs [hunk 1]
```diff
@@ -208,15 +227,20 @@ impl SingleByteDecoder {
...
                             }
+                            // Safety: We are now at `converted <= length`. We should *not* `continue`
+                            // the loop without reverifying
                             converted += 1;
                             if b < 60 {
                                 // We've got punctuation
                                 if converted == length {
                                     return (pending, length, length);
                                 }
+                                // Safety: we're back to `converted <= length` because of the == above
```
Comment (riking) [inline src/single_byte.rs:241]: we're back to `converted < length`

File: src/x_user_defined.rs [hunk 2]
```diff
@@ -116,10 +116,15 @@ impl UserDefinedDecoder {
...
         let dst_ptr = dst.as_mut_ptr();
+        // Safety: This is `for i in 0..length / 16`
         for i in 0..simd_iterations {
+            // Safety: This is in bounds: length is the minumum valid length for both src/dst
+            // and i ranges to length/16, so multiplying by 16 will always be `< length` and can do
+            // a 16 byte read
             let input = unsafe { load16_unaligned(src_ptr.add(i * 16)) };
             let (first, second) = simd_unpack(input);
             unsafe {
+                // Safety: same as above, but this is two consecutive 8-byte reads
```
Comment (riking) [inline src/x_user_defined.rs:127]: two writes

**Review by hsivonen — APPROVED** [review]
Thank you!

**Pull request merged**